crypto/rand/rand_vxworks.c

```diff
@@ -0,0 +1,172 @@
+# include "rand_lcl.h"
+# include "internal/rand_int.h"
+# include "internal/cryptlib.h"
+
+# if defined(OPENSSL_RAND_SEED_OS) 
+#  if _WRS_VXWORKS_MAJOR >= 7
+#    define RAND_SEED_VXRANDLIB
+#  else
+#    error "VxWorks <7 only support RAND_SEED_NONE"
+#  endif
+# endif
+
+# if defined(RAND_SEED_VXRANDLIB)    
+#  include <randomNumGen.h>
+# endif
+
+/* Macro to convert two thirty two bit values into a sixty four bit one */
+# define TWO32TO64(a, b) ((((uint64_t)(a)) << 32) + (b))
+
+static uint64_t get_time_stamp(void)
+{
+    struct timespec ts;
+
+    if (clock_gettime(CLOCK_REALTIME, &ts) == 0)
+        return TWO32TO64(ts.tv_sec, ts.tv_nsec);
+    return time(NULL);
+}
+
+static uint64_t get_timer_bits(void)
+{
+    uint64_t res = OPENSSL_rdtsc();
+
+    if (res != 0)
+        return res;
+
+    {
+        struct timespec ts;
+
+        if (clock_gettime(CLOCK_MONOTONIC, &ts) == 0)
+            return TWO32TO64(ts.tv_sec, ts.tv_nsec);
+    }
+    return time(NULL);
+}
+
+/* 
+ *	empty implementation 
+ *  vxworks does not need to init/cleanup or keep open the random lib
+ */
+int rand_pool_init(void)
+{
+    return 1;
+}
+
+void rand_pool_cleanup(void)
+{
+}
+
+void rand_pool_keep_random_devices_open(int keep)
+{
+}
+
+int rand_pool_add_additional_data(RAND_POOL *pool)
+{
+    struct {
+        CRYPTO_THREAD_ID tid;
+        uint64_t time;
+    } data = { 0 };
+
+    /*
+     * Add some noise from the thread id and a high resolution timer.
+     * The thread id adds a little randomness if the drbg is accessed
+     * concurrently (which is the case for the <master> drbg).
+     */
+    data.tid = CRYPTO_THREAD_get_current_id();
+    data.time = get_timer_bits();
+
+    return rand_pool_add(pool, (unsigned char *)&data, sizeof(data), 0);
+}
+
+int rand_pool_add_nonce_data(RAND_POOL *pool)
+{
+    struct {
+        pid_t pid;
+        CRYPTO_THREAD_ID tid;
+        uint64_t time;
+    } data = { 0 };
+
+    /*
+     * Add process id, thread id, and a high resolution timestamp to
+     * ensure that the nonce is unique whith high probability for
+     * different process instances.
+     */
+    data.pid = getpid();
+    data.tid = CRYPTO_THREAD_get_current_id();
+
+    data.time = get_time_stamp();
+
+    return rand_pool_add(pool, (unsigned char *)&data, sizeof(data), 0);
+}
+# if defined(OPENSSL_RAND_SEED_NONE)
+size_t rand_pool_acquire_entropy(RAND_POOL *pool)
+{
+    short int code;
+    int i, k;
+    size_t bytes_needed;
+    struct timespec t1;
+    struct timespec t2;
+    unsigned char v;
+    long long duration;
+    extern void s$sleep2(long long *_duration, short int *_code);
+
+    bytes_needed = rand_pool_bytes_needed(pool, 4 /*entropy_factor*/);
+
+    for (i = 0; i < bytes_needed; i++) {
+        /* wait until monotonic clock has a new value */
+        clock_gettime(CLOCK_MONOTONIC, &t1);
+        do {
+            clock_gettime(CLOCK_MONOTONIC, &t2);
+        } while (t1.tv_sec == t2.tv_sec && t1.tv_nsec == t2.tv_nsec);

```

Comment:
please merge the count how many times the loop executed into the rand_pool

Reply:
> Out of curiosity, how does the randlib behave, short after system boot (on a bare metal hardware)?
> Does a RAND_bytes behave, does it fail, or succeed, or wait until there is entropy?

i tested it and it gives me quite some random values directly after boot before it returns error. i tested with this loop:

void randTest()
{
	unsigned char buf[10];
	int i,j;
	for (i=0; i<1000; i++)
	{
		STATUS retVal;
		retVal=randBytes(buf, 10);
		if (retVal == OK )
		{
			printf ("Got 10 bytes of randomness: ");
			for (j=0; j<10; j++)
			{
				printf ("%d ", buf[j]);
			}
			printf ("\n");
		} else {
			printf ("Error getting 10 random bytes...\n");
		}
	}
}

which is testing it the hard time (no waits at all). It results in ~60x10 bytes beeing deliverd w/o errors and then some errors followed by another block of randomness and so on...

If i wait a bit after an error the randomness will come back at times.

@bernd-edlinger : would it make sense to retry in cases of not OK until it works or is it better to pass the erorr on?

Reply:
If we let it fail here, the chances are that the caller of RAND_bytes does not check the return
code, or it does at least not have no good strategy to handle the error.
I would prefer to wait, unless it is a permanent error.